diff --git a/contrib/test_disk-server/CMakeLists.txt b/contrib/test_disk-server/CMakeLists.txt
index 271c60b05c..3ce451587f 100644
--- a/contrib/test_disk-server/CMakeLists.txt
+++ b/contrib/test_disk-server/CMakeLists.txt
@@ -8,4 +8,25 @@ add_executable(${PROJECT_NAME})
 target_sources(${PROJECT_NAME}
     PRIVATE
         server.cpp
+        server_toolkit/server_toolkit.cpp
+)
+
+target_include_directories(${PROJECT_NAME}
+    PUBLIC
+        ${PROJECT_SOURCE_DIR}
+)
+
+# ==============================================================
+
+add_executable(${PROJECT_NAME}_change_parameters)
+
+target_sources(${PROJECT_NAME}_change_parameters
+    PRIVATE
+        change_parameters.cpp
+        server_toolkit/server_toolkit.cpp
+)
+
+target_include_directories(${PROJECT_NAME}
+        PUBLIC
+        ${PROJECT_SOURCE_DIR}
 )
diff --git a/contrib/test_disk-server/change_parameters.cpp b/contrib/test_disk-server/change_parameters.cpp
new file mode 100644
index 0000000000..a6ba9efd1d
--- /dev/null
+++ b/contrib/test_disk-server/change_parameters.cpp
@@ -0,0 +1,5 @@
+#include "server_toolkit/server_toolkit.h"
+
+int main() {
+    change_server_parameters("127.0.0.1", 31234, 10, 40);
+}
\ No newline at end of file
diff --git a/contrib/test_disk-server/server.cpp b/contrib/test_disk-server/server.cpp
index a7a4148459..2e095a0389 100644
--- a/contrib/test_disk-server/server.cpp
+++ b/contrib/test_disk-server/server.cpp
@@ -1,63 +1,34 @@
-#include <arpa/inet.h>
-#include <cstring>
-#include <format>
-#include <iomanip>
-#include <iostream>
-#include <netinet/in.h>
-#include <string>
-#include <bits/this_thread_sleep.h>
-#include <sys/socket.h>
-
-#include "structs.h"
-
+#include "server_toolkit/server_toolkit.h"
 
 int main() {
-    sockaddr_in addr = {
-        .sin_family = AF_INET,
-        .sin_port = htons(31234),
-        .sin_zero = 0
-    };
-
-    inet_aton("127.0.0.1", &addr.sin_addr);
-
-    int fd = socket(AF_INET, SOCK_STREAM, 0);
-
-    int opt = 1;
-    setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));
-
-    bind(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr));
-
-    listen(fd, 1);
+    int server_fd = create_server("127.0.0.1", 31234);
+    if (server_fd == -1) {
+        return -1;
+    }
 
-    uint32_t read_latency_ms = 5;
-    uint32_t write_latency_ms = 5;
+    std::atomic<uint32_t> read_latency_ms = 34;
+    std::atomic<uint32_t> write_latency_ms = 7;
 
-    while (true) {
-        int client_fd = accept(fd, nullptr, nullptr);
-        std::cout << std::format("---- New connection ---- ") << std::endl;
+    run_server(server_fd, [&](const ServerRequest& server_req) {
+        ServerResponse server_res{};
 
-        while (true) {
-            ServerRequest server_req;
-            ssize_t bytes = recv(client_fd, &server_req, sizeof(server_req), 0);
-            if (bytes <= 0) {
+        switch (server_req.source) {
+            case SOURCE_CHANGE_PARAMETERS:
+                read_latency_ms = server_req.change_parameters_payload.read_latency_ms;
+                write_latency_ms = server_req.change_parameters_payload.write_latency_ms;
                 break;
-            }
+            case SOURCE_TEST_DISK:
+                std::cout << std::format("Command: 0x{:02x}", server_req.test_disk_payload.command) << std::endl;
 
-            if (server_req.source != SOURCE_TEST_DISK) {
-                continue;
-            }
-
-            std::cout << std::format("Received: id = {}, command = 0x{:02x}", server_req.test_disk_payload.disk_id, static_cast<int>(server_req.test_disk_payload.command)) << std::endl;
-
-            ServerResponse server_res;
-            server_res.latency = server_req.test_disk_payload.command == 0x28 || server_req.test_disk_payload.command == 0x08 ? read_latency_ms : write_latency_ms;
-
-            send(client_fd, &server_res, sizeof(server_res), 0);
+                server_res.latency =
+                    server_req.test_disk_payload.command == 0x28 || server_req.test_disk_payload.command == 0x08
+                    ? read_latency_ms
+                    : write_latency_ms;
+                break;
+        }
 
-            // std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        // std::this_thread::sleep_for(std::chrono::milliseconds(150));
 
-            std::cout << "Sent: latency = " << server_res.latency << std::endl;
-        }
-        std::cout << "---- Connection closed ----" << std::endl;
-    }
+        return server_res;
+    });
 }
diff --git a/contrib/test_disk-server/server_toolkit/server_toolkit.cpp b/contrib/test_disk-server/server_toolkit/server_toolkit.cpp
new file mode 100644
index 0000000000..a58ce334ee
--- /dev/null
+++ b/contrib/test_disk-server/server_toolkit/server_toolkit.cpp
@@ -0,0 +1,132 @@
+#include "server_toolkit.h"
+
+int create_server(const std::string& ip, uint16_t port) {
+    sockaddr_in addr = {
+        .sin_family = AF_INET,
+        .sin_port = htons(port),
+        .sin_zero = 0
+    };
+
+    if (inet_aton(ip.c_str(), &addr.sin_addr) == 0) {
+        std::cerr << "Invalid address" << std::endl;
+        return -1;
+    }
+
+    int fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (fd == -1) {
+        std::cerr << "Failed to create socket: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    int opt = 1;
+    if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) == -1) {
+        close(fd);
+        std::cerr << "Unable to set SO_REUSEADDR option: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    if (bind(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) == -1) {
+        close(fd);
+        std::cerr << "Unable to bind socket: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    return fd;
+}
+
+// ============================================================================
+
+int run_server(int server_fd, const Callback& callback) {
+    if (listen(server_fd, 5) == -1) {
+        close(server_fd);
+        std::cerr << "Unable to listen server socket: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    while (true) {
+        int client_fd = accept(server_fd, nullptr, nullptr);
+        if (client_fd == -1) {
+            std::cerr << "Unable to accept client" << std::endl;
+            continue;
+        }
+
+        std::thread([client_fd, callback]() {
+            client_routine(client_fd, callback);
+        }).detach();
+    }
+
+    return 0;
+}
+
+// ============================================================================
+
+void client_routine(int client_fd, const Callback& callback) {
+    std::cout << "---- New connection ---- " << std::endl;
+
+    while (true) {
+        ServerRequest server_req;
+        ssize_t bytes = recv(client_fd, &server_req, sizeof(server_req), 0);
+        if (bytes <= 0) {
+            break;
+        }
+
+        ServerResponse server_res = callback(server_req);
+
+        if (send(client_fd, &server_res, sizeof(server_res), 0) == -1) {
+            std::cerr << "Unable to send response to client socket: " << strerror(errno) << std::endl;
+        } else {
+            std::cout << "Sent: latency = " << server_res.latency << std::endl;
+        }
+    }
+
+    std::cout << "---- Connection closed ----" << std::endl;
+
+    close(client_fd);
+}
+
+// ============================================================================
+
+int change_server_parameters(const std::string& ip, uint16_t port, uint32_t read_latency_ms, uint32_t write_latency_ms) {
+    // server address
+    sockaddr_in addr = {
+        .sin_family = AF_INET,
+        .sin_port = htons(port),
+        .sin_zero = 0
+    };
+
+    if (inet_aton(ip.c_str(), &addr.sin_addr) == 0) {
+        std::cerr << "Invalid address" << std::endl;
+        return -1;
+    }
+
+    // connect to server
+    int fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (fd == -1) {
+        std::cerr << "Failed to create socket: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    if (connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) == -1) {
+        close(fd);
+        std::cerr << "Unable to connect to server: " << strerror(errno) << std::endl;
+        return -1;
+    }
+
+    // send request
+    ServerRequest server_req = {
+        .source = SOURCE_CHANGE_PARAMETERS,
+        .change_parameters_payload = {
+            .read_latency_ms = read_latency_ms,
+            .write_latency_ms = write_latency_ms,
+        }
+    };
+
+    if (send(fd, &server_req, sizeof(server_req), 0) == -1) {
+        std::cerr << "Unable to send request to client socket: " << strerror(errno) << std::endl;
+    } else {
+        std::cout << "Sent: read_latency_ms = " << read_latency_ms << ", write_latency_ms = " << write_latency_ms << std::endl;
+    }
+
+    close(fd);
+    return 0;
+}
diff --git a/contrib/test_disk-server/server_toolkit/server_toolkit.h b/contrib/test_disk-server/server_toolkit/server_toolkit.h
new file mode 100644
index 0000000000..6d9845f770
--- /dev/null
+++ b/contrib/test_disk-server/server_toolkit/server_toolkit.h
@@ -0,0 +1,28 @@
+#ifndef SERVER_H
+#define SERVER_H
+
+#include <arpa/inet.h>
+#include <cstring>
+#include <format>
+#include <functional>
+#include <iomanip>
+#include <iostream>
+#include <netinet/in.h>
+#include <string>
+#include <unistd.h>
+#include <bits/this_thread_sleep.h>
+#include <sys/socket.h>
+#include <thread>
+
+#include "structs.h"
+
+int create_server(const std::string& ip, uint16_t port);
+
+using Callback = std::function<ServerResponse(ServerRequest)>;
+int run_server(int server_fd, const Callback& callback);
+
+void client_routine(int client_fd, const Callback& callback);
+
+int change_server_parameters(const std::string& ip, uint16_t port, uint32_t read_latency_ms, uint32_t write_latency_ms);
+
+#endif //SERVER_H
